### ISSUE TITLE
Set up funnel with plan info

### ISSUE THREAD

**Issue by Saltzshaker** [popuparchive/pop-up-archive#1298]
Notes here: https://docs.google.com/document/d/1iXFpu84AqQQ2MqdGwxizIf1EUYVqwv86e_pif8zLGcE/edit?usp=sharing

Set up Mixpanel funnel for 'Registered' (demo) to 'Upload' to 'Subscription Change'. https://mixpanel.com/report/397857/funnels/#view/1438489/from_date:-89,to_date:0
Subscription change is showing up as 0 even though plans have changed. Contacting Mixpanel for support. 

Goal for Joan: Have working funnel for 'Registered' (demo) to 'Upload' to 'Clicked Upgrade' to 'Subscription Change' (upgrade)

**Comment by Saltzshaker** [comment]
Fixing this funnel requires rewriting how we assign distinct_ids for these events, both in Ruby and JS. Assigning the ticket to @baileyspace for this step. She'll reassign to me (probs for Mencken milestone) to test and set up in Mixpanel funnel UI. 

In terms of how to rewrite the events, Mixpanel support says:
"If you are using both our Javascript and Ruby libraries, the key step to here will be to alias your internal user id to the original Mixpanel-assigned distinct_id upon registration. From reading your helpful description and looking through your project, my hypothesis about your 'Registration' event is that you are experiencing a classic case of a client/server aliasing race condition. 

In short: if you send events to the new ID and the alias is not ready yet, those events won't ever "connect" with the original anonymous ID. There are two solutions, both of which involve removing the ambiguity which causes the race condition. **These are the options:** 1) implement a small delay (~500ms) between your Alias call and any server-side events that are sent to the new ID. This gives the backend time to process the alias, so that all events will be redirected properly. 2) The safest option: simply grab the anonymous distinct_id from the client-side (you can parse the cookie like you've done above, or use mixpanel.get_distinct_id() from within the "loaded" callback of the mixpanel.init() function [see http://bit.ly/1I7viOF]) and use that original ID as the distinct_id of any events you are sending immediately upon registration. You can start using your internal user IDs anytime after that, as long as ~500ms has passed and you can be sure that the alias is complete.

**Comment by Saltzshaker** [comment]
Registered, subscription change, and audio upload complete are all appearing with data in the same funnel with consistent customer info. However, "subscription change" event still doesn't appear to be logging subscription changes accurately. 

I signed up as emily+mixpanel2@popuparchive.com. Before changing plans, Mixpanel logged a Registration and Subscription change event, both with premium community. When I actually did change plans several times, no more subscription changes were logged. Assigning back to @baileyspace

**Comment by Saltzshaker** [comment]
Subscription change is logged for every subscription change after registration. Registration, subscription change, and audio upload are all working in the same funnel. Marking ready to deploy

@baileyspace also made a change to update "Plan" field upon subscription change. In order to update past users before this change goes into effect, run the following rake task upon deploy to production: rake users:mixpanel_plan

Note: we also encountered an edge case for multiple users who have the same cookies (e.g. me creating emily+mixpanel3@popuparchive.com and emily+mixpanel4@popuparchive.com). If this happens the user info for the first user (emily+mixpanel3@popuparchive.com) got overwritten by the second (emily+mixpanel4@popuparchive.com). Since most users should not have multiple accounts with the same cookies, @baileyspace and I decided it's not worth troubleshooting further. Documenting here for future reference.
